### PR TITLE
[Merged by Bors] - refactor: move internals from `entity_ref` to `World`, add `SAFETY` comments

### PR DESCRIPTION
**Pull request by jakobhellermann**
# Objective

There are some utility functions for actually working with `Storages` inside `entity_ref.rs` that are used both for `EntityRef/EntityMut` and `World`, with a `// TODO: move to Storages`.
This PR moves them to private methods on `World`, because that's the safest API boundary. On `Storages` you would need to ensure that you pass `Components` from the same world.

## Solution

- move get_component[_with_type], get_ticks[_with_type], get_component_and_ticks[_with_type] to `World` (still pub(crate))
- replace `pub use entity_ref::*;` with `pub use entity_ref::{EntityRef, EntityMut}` and qualified `entity_ref::get_mut[_by_id]` in `world.rs`
- add safety comments to a bunch of methods

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
I agree with the code reorg. Lemme take a look at the safety comments...

File: crates/bevy_ecs/src/storage/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:33]: We're passing in a `type_id` here, so I don't think this applies.

File: crates/bevy_ecs/src/storage/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:32]: A specific archetype isn't given, so this comment is unclear.

File: crates/bevy_ecs/src/storage/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:52]: Ditto for "given archetype". This seems to be a problem on pretty much every method here :)

File: crates/bevy_ecs/src/storage/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:73]: It's not clear what `T` refers to in the context of this function.

File: crates/bevy_ecs/src/storage/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:53]: I'd like a bit more advice here on what it means for a component id to be valid. Probably best suited to living on the `ComponentId` docs though.

File: crates/bevy_ecs/src/storage/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/mod.rs:188]: T is not defined here.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Okay, so the safety docs have some problems 😂 They're pretty much all repeated, and I didn't call them out in each location to reduce spam.

Not your fault by any means (these docs are much better than we had), but eh, I'd like to clean them up while we're here.

File: crates/bevy_ecs/src/world/entity_ref.rs
Comment (james7132) [inline crates/bevy_ecs/src/world/entity_ref.rs:242]: Almost all of these are using borrows from `World`. Wouldn't it make more sense to move these to `World` than `Storages`?

File: crates/bevy_ecs/src/storage/mod.rs
Comment (james7132) [inline crates/bevy_ecs/src/storage/mod.rs:37]: These were `pub(crate)` before, why are they now just `pub`?

File: crates/bevy_ecs/src/world/entity_ref.rs
Comment (james7132) [reply crates/bevy_ecs/src/world/entity_ref.rs:243]: If it cannot be safely used without them, it's best to put it on the interface where it's most cleanly used and has fewer safety invariants.

File: crates/bevy_ecs/src/storage/mod.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/mod.rs:36]: I'd err on the side of caution here. Unless we have a solid external use case, having more `pub`, particularly for items that mutably change the state of the `World`, is just asking for public facing unsoundness.

File: crates/bevy_ecs/src/world/entity_ref.rs
Comment (james7132) [inline crates/bevy_ecs/src/world/entity_ref.rs:667]:
```suggestion
```

This is redundant with the above.

File: crates/bevy_ecs/src/world/mod.rs
Comment (james7132) [inline crates/bevy_ecs/src/world/mod.rs:1699]: Any reason why these couldn't be in the block above?

**Review by james7132 — CHANGES_REQUESTED** [review]

**Review by james7132 — APPROVED** [review]

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1698]: This needs safety comments.

File: crates/bevy_ecs/src/world/entity_ref.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_ref.rs:895]: The PR description suggests that this is a method on Storages (which I like better), but here it's a freestanding function.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Two concerns, but once those are fixed up I'm happy to merge this. It's better organized, and the safety comments are much better.

**Review by alice-i-cecile — APPROVED** [review]
LGTM now once the PR description matches where the methods were moved, one way or another.

**Pull request closed without merging**
